---
output: 
  word_document:
    reference_docx: rMarkdown-word-styles-reference-01.docx
title: "Sexually antagonistic coevolution between the sex chromosomes of *Drosophila melanogaster*"
author:
  - Katrine K. Lund-Hansen (*,1,2)
  - Colin Olito (1)
  - Edward H. Morrow (2)
  - Jessica K. Abbott (1)
abstract: "(1) Department of Biology, Section for Evolutionary Ecology, Lund University, Lund 223 62, Sweden. (2) School of Life Sciences, University of Sussex, Brighton BN1 9QG, United Kingdom. *Corresponding author e-mail*: Katrine.Lund-Hansen@biol.lu.      *Manuscript elements*: Figure 1, Figure 2, Figure 3;  *Online Supplementary Material*: Appendix A – Methods: Figure S1, Table S1, Table S2, Table S3, Table S4. *Keywords*: Coevolution; Compensatory evolution; Interlocus sexual conflict; Sexual dimorphism; Sex chromosome evolution. *Manuscript Type*: Major Article"
date: "`r format(Sys.time(), '%B %d, %Y')`"
---

# Abstract -- No more than $250$ words

Antagonistic interactions between the sexes have been proposed to be important drivers of evolutionary divergence. This form of interlocus sexual conflict is modelled as a conflict between alleles at two interacting loci, whose identity remain largely unknown. Here we suggest that if these two loci are located on the sex chromosomes this can lead to a perpetual cycle of antagonistic coevolution between the sex chromosomes and the two sexes. To test this hypothesis we experimentally interrupted the cycle of coevolution by exchanging the sex chromosomes between five wild type populations of Drosophila melanogaster and measured the consequences of these chromosomal replacements for a range of phenotypic and life history traits. We found that disrupting the cycle of coevolution increased male fitness significantly for 10 out of 20 novel genotypes. The benefit in male fitness came at a cost to overall female fitness, because the novel genotypes negatively affected the viability of offspring egg-to-adult survival, suggesting sexually antagonistic effects drive these results. After 25 generations of experimental evolution, the initial increase in relative male fitness disappeared, and there were no longer any significant differences between novel genotypes and their wild-type counterparts. These results add further support to the idea that a cycle of antagonistic coevolution can occur between sex chromosomes, and illustrates how each independent cycle can increase genetic and phenotypic divergence between populations of D. melanogaster.

# Introduction

In species with separate sexes, conflict can impact the shared genome when the evolutionary interests of each sex, mostly due to anisogamy, are in opposition to each other. Sexual conflict is an evolutionary dynamic that can prevent each sex from reaching individual fitness optima, and can be particularly harmful when alleles beneficial to one sex are simultaneously disadvantageous to the other, e.g. sexual antagonism (1, 2). It has previously been theorised that an intergenomic conflict could occur between different alleles in different individuals, as a change in sexually antagonistic variance in the direction of either sex (i.e. an accumulation of either male or female beneficial sexual antagonistic alleles) would push the other sex further away from its optimum fitness value (i.e. increased gender-load) (3, 4). As each sex has a pool of genes with sex-limited expression, these could be used to ameliorate the increased gender-load and thereby counteract the negative effects of the sexually antagonistic genes, and so led to antagonistic coevolution (4). We hypothesised that an intergenomic conflict would be greater between the sex chromosomes due to their sex-limited mode of inheritance, their specialised sex-limited gene expression, and sex determination (5) and would consequently have a higher degree of opposing interests than the autosomes, which are equally shared between the two sexes. If there is strong intergenomic conflict between the sex chromosomes this could lead to antagonistic coevolution of the sex chromosomes. Previously, the standard view of the Y-chromosome has been of a highly heterochromatic degraded chromosome with no genomic agency. However, in recent years, multiple studies in Drosophilae have shown an effect of the Y-chromosome on a broader range of phenotypic and life-history traits, such as male fitness (6), geotaxis (7), male courtship song (8), thermal adaptation (9), suppression of the sex-ratio gene on the X-chromosome (10, 11) and male lifespan (12). These results indicate that variation across different Y-chromosomes can have far-reaching consequences for diverse complex and life history traits. This is apparently not due to genetic variation, since Zurovcova and Eanes did not observe any polymorphisms in a protein-coding gene between 11 different D. melanogaster Y-chromosomes (13). Instead, recent data indicates that this between-Y-chromosome variance could be due to epigenetic effects of the chromosome (14). Lemos et al. showed that the Y-chromosome affected the expression of many non Y-linked genes, and a majority of these genes had a male-biased expression (15). Since then, Y-linked regulatory variation has been shown to control the expression of immunity- and mitochondria-related genes (14) as well as testes-specific genes (16), and silencing of X-linked rDNA in males (17). Furthermore, the genomic background is also apparently important for the epigenetic function of Y-chromosome (18). Together, these results indicate that the epigenetic interaction between the Y-chromosome and the rest of the genome is important for genome-wide patterns of gene expression, especially genes important for male fitness.

To empirically test if the sex chromosomes coevolve in D. melanogaster, we designed a round-robin crossing scheme where any interactions between the sex chromosomes would be disrupted by exchanging sex chromosomes between five wild-type populations. It is expected that sex chromosome coevolution would follow different trajectories in different populations due to random mutations and the interaction between the environment and sexual conflict (19). By choosing wild-type populations from all over the world with different life histories and adaptations to different climatic regimes, we would expect the sex chromosomes to have followed different paths of coevolution. To ensure we captured as much variation as possible, we selected the five wild-type populations from large, outbred laboratory fly stocks, which should ensure low levels of inbreeding. The round-robin crossing scheme was designed so that one cross between two populations created two novel genotypes where the Y-chromosomes from each population were put into the genetic background of the other population (hereafter: novel Y), and a further two novel genotypes where the X-chromosomes were swapped (hereafter: novel X), thus generating all combinations of the novel sex chromosome pairs and autosomal backgrounds. To confirm that any effect we found was due to changing a sex chromosome, we also created novel XY genotypes where both of the sex chromosomes from the same population were placed alongside the autosomes from another population. If the sex chromosomes coevolve via a process by which variation on the Y-chromosome increases male fitness, and variation on the X-chromosome responds by ameliorating these, then removing either of the sex chromosomes from a pair would release the Y-chromosome from any counter adaptations on the X-chromosome, and should increase male fitness. We therefore used male reproductive fitness (the number of live offspring sired by a male) as an estimate of overall male fitness, to confirm if the novel genotypes had disrupted the coevolution between the sex chromosomes.

As we outline below in the *Results and Discussion* section, our experimental results indicate that disrupting coevolved sex chromosome pairs resulted in direct fitness benefits to males *via* improved fertilization success under sperm competition, and fitness costs to both males and females *via* reduced offspring survival. Moreover, these antagonstic fitness effects were ephemeral -- they were not observed after 25 generations of experimental evolution in populations with novel X or Y genotypes -- suggesting a possible role for compensatory evolution driven by selection on females to reduce the fitness cost of reduced offspring survival. To formalize these hypotheses and assess their plausibility, we developed population genetic models describing the evolution of two interacting biallelic loci located in different genomic regions where parental genotypes determine offspring survival: a non-recombining Y-linked locus influencing both adult male mating success (i.e., sperm competition) and subsequent offspring survival: and a compensatory autosomal or X-linked locus influencing offspring survival only. 


# Results and Discussion

## **Model results/discussion**
We present two population genetic models, identified by the location of the compensatory locus: the Autosomal and X-linked models respectively. Our theoretical analyses focus on (i) evolutionary invasion of single mutants at the Y-linked and compensatory locus; and (ii) single bouts of coevolution beginning with invasion of a single-copy mutant $y$ chromosome in a population initially fixed for the wild-type $Y$ chromosome and for the $A$ (or $X$) allele at the compensatory locus, and completes when both mutant alleles $y$ and $a$ (or $x$) become fixed in the population (see Connallon et al. 2017 for a similar approach in the context of mito-nuclear coevolution). In both models (Autosomal or X-linked), the mutant $y$ chromosome that increases male fertilization success of $1 + s_m$ relative to wild type males (denoted $Y$), and reduced survival of offspring produced by matings with females carrying the wild-type compensatory allele, $a$ (or $x$); while females carrying the mutant $a$ allele may incur a ’cost of compensation’ caused by reduced offspring survival when mating with wild-type ($Y$) males (see ***Table X for a full summary of fitness expressions***). These fitness expressions create a scenario similar to standard theories of compensatory evolution (e.g., **Kimura 1985, Weinrich \& Chao 2005**) where each of the mutant alleles ($y$ and either $a$ or $x$) is deleterious for offspring survival in isolation, with compensation (i.e., no reduction in offspring survival) requiring that the other parent has the appropriate mutant genotype at the other locus (a full description of the models is presented in the *Methods* section and in *Appendix **X** of the Online Supplementary Material*).

Evolutionary invasion analyses of our models reveales three key results. First, as expected the invasion of a rare mutant $y$ chromosome requires that the net fitness effect through both male mating success and offspring survival is male-beneficial, regardless of the genomic location of the compensatory locus. Specifically, invasion of $y$ requires that $\delta = s_m - s_c > 0$. 


To empirically test if the sex chromosomes in D. melanogaster coevolve we created 20 novel genotypes where either the X-chromosome (novel X) or the Y-chromosome (novel Y) from one wild-type population was incorporated into another population (SI Appendix, Fig. S1). This was carried out as a round-robin cross between five outbred laboratory adapted wild-type populations derived from four continents and three climatic zones (Fig. 1). We found a significant effect of genotypes on male relative fitness ($P = 2.49e-04$, Fig. 2A, SI Appendix, Table S1), with males from the novel X or novel Y genotypes having a significant higher relative fitness than the wild-type males (Turkey HSD: Novel X–wild-type: $P = 2.08e-04$; Novel Y–wild-type: $P = 2.50e-03$). To exclude interactions between the novel sex chromosome and the autosomes, we also created novel XY genotypes where a pair of sex chromosomes from one population were introduced into another population (SI Appendix, Fig. S2). We found no significant effect of genotype on relative fitness in the assay with novel XY genotypes ($P = 0.98$, Fig. 2B, SI Appendix, Table S1). It therefore seems that introducing a new sex chromosome resulted in the difference in fitness, and was not influenced by the autosome background. To investigate this in more details, we calculated the difference in fitness between the novel genotypes and their wild-type counterparts ($\Delta$ Fitness = $\omega$ novel genotype - $\omega$ wild type). We use bootstrap to model the data, and found that 10 of the 20 novel genotypes were significantly different from zero in a positive direction (Fig. 3, SI Appendix, Table S2). Interestingly, 16 out of the 20 point estimates were positive, which is significant from what would be expected ($P = 0.01$). So, with half of the novel genotype having confidence intervals not overlapping zero and a significant majority being positive, the data shows that, in general, changing a sex chromosome has a significant positive effect on male fitness. This was consistent with our predictions, as we expected to see an increase in male fitness if the interactions between the sex chromosomes were antagonistic in nature.

To investigate which male-beneficial functions could contribute to the increased male fitness found in the novel genotypes, other than male fitness, we looked at phenotypic traits that have been shown to be correlated with male fitness, such as size (20, 21), courtship rate (20), and sperm competition (22). Body size is an important factor in male fitness (21), and larger males have been shown to have a mating advantage (23), so we measured thorax length of the novel males as an estimate for body size. We found that there was a significant difference in size between the novel genotypes and wild-type males ($P < 2.2e-16$, SI Appendix, Fig.S3.A and Table S1). In fact, novel X males were significantly larger than both wild-type (Turkey HSD: $P = 3.46e-11$) and novel Y males (Turkey HSD: $P < 0.01$), whereas there were no significant difference between novel Y and wild-type males (Turkey HSD: $P = 0.13$). It is therefore possible that a larger body size of novel X males was a factor in the increase fitness. This did not seem to be the case for novel Y males, and so it suggest that different genetic mechanisms may be responsible for the increased male fitness exhibited by the two novel genotypes. Another phenotypic trait that could explain the increase in fitness is courtship activity. Even though we found a significant difference in courtship rate between the different genotypes ($P = 1.83e-04$, SI Appendix, Fig.S3.B and Table S1), this was due to a significant higher courtship rate in wild-type males than males from novel genotypes (Turkey HSD: Novel X–wild-type: $P = 1.04e-04$, Novel Y–wild-type: $P = 0.03$). Unlike previous experiments which showed that larger males have a higher courtship and remating rate (20, 21), we observed that males from novel X were less active during courtship. This conflicts with the fitness results as we expect that high fitness males would engage and participate in more matings than low fitness males, but Bedhomme et al. also found lower courtship rate in high fitness males (24). A possible explanation for the lower courtship rate could be that males become more efficient at courtship thereby decreasing courtship duration or perhaps they outperform other males though sperm competition, and sires a larger percentage of a female’s offspring. We found a significant effect of genotype on the ability to replace sperm ($P = 0.03$, SI Appendix, Fig.S3.C and Table S1). Unlike previous studies we did not find that larger males preformed better in sperm competition (25) (Turkey HSD: Novel X–wild-type: $P = 0.12$), instead we found that novel Y males were significantly better at sperm competition than wild-type males (Turkey HSD: Novel Y–wild-type: $P = 0.03$).

If the interaction between the sex chromosomes were antagonistic, we would expect to find a decrease in female fitness. For example, male accessory gland secretions have been found to induce and/or increase female egg production (26, 27). It is therefore possible that the higher relative fitness of novel genotype males is due to production of larger or more potent accessory gland secretions, inducing high levels of ovulation and thus causing greater harm. However, we did not find any significant difference between the three genotypes in inducing egg-laying in females ($P = 0.18$, SI Appendix, Fig. S3.D and Table S1). Another possibility is that female lifespan is compromised following mating with novel genotype males. It has previously been shown that mating with males have a negative effect on female lifespan (28–30), and though the results confirmed a reduction in female lifespan when continuously exposed to males, we did not find any significant difference between the genotypes ($P = 0.1$, SI Appendix, Fig. S4.A and Table S3).

Although we could not find any evidence of direct costs to females, it is possible that males would trade-off elevated reproductive success over other life-history traits. One such trade-off that has been well studied is between reproduction and longevity in males (31), which is also interesting in this context as the Y-chromosome also has been shown to affect male lifespan (12). However, we did not find any difference in lifespan between the three genotypes ($P = 1$, SI Appendix Fig. S4.B and Table S3). Prowse and Partridge have shown that virgin males live significantly longer than males who mate throughout life, and so it is possible that the experimental setup, where the males were housed without females, was too ineffective to separate any differences between the treatments (32).

Another possibility is that there is a trade-off between number of offspring and offspring quality. Friberg et al. found evidence of sex-linked paternal effect on egg-to-adult survival (33). Though we found no difference in number of eggs laid by females mated with the novel genotypes (SI Appendix, Fig. S3.D), we did find a significant effect of genotype on total offspring number ($P = 0.02$, SI Appendix, Fig. S5.A and Table S1). This lower number of adult offspring could be due to decreased egg-to-adult viability. In line with this hypothesis, we did find a significant difference in egg-to-adult viability ($P = 7.36e-03$, SI Appendix, Fig. S5.B and Table S1), with novel X males siring significantly lower numbers of viable offspring than the other two genotypes (Turkey HSD: Novel X–novel Y: $P = 8.72e-03$; Novel X–wild-type: $P = 0.04$). This is consistent with results showing that offspring sired by larger males have a lower survival rate (30). Since there are no known sex ratio meiotic drivers in D. melanogaster (34), and we found no significant difference in sex ratio ($P = 0.22$, SI Appendix, Fig. S5.C, Table S1), decreased offspring survival is unlikely to be explained by meiotic drive. We were not able to establish which harmful effects novel Y males had on female fitness through the assays performed in this experiment. A possible cost to female fitness not investigated in this experiment is the reduction of female willingness to remate after mating with a novel Y male, which could lead to a reduction in lifetime reproductive fitness of females.

To complete the empirical test of coevolution between the sex chromosomes in D. melanogaster, we first disrupted the coevolution by introducing a novel sex chromosome and then allowed time for the novel pairs to co-adapt during an evolution experiment. If the coevolution between the sex chromosomes were antagonistic, disruption of the interaction would be harmful to one sex but beneficial to the other. This perturbed state should then settle in time and the novel genotypes should return to their original state, which is what we observed after 25 generations. Taken together, all of these results indicate that sex chromosomes do coevolve in an antagonistic manner.

After 25 generations of evolution experiment (SI Appendix, Fig. S6), we repeated the male reproductive fitness assay in order to test how the interaction between the novel pairs of sex chromosomes had been affected during this period. We found that there was no longer a significant effect of treatment on male fitness ($P = 0.85$, Fig. 4.A, SI Appendix, Table S4). Indeed, we found that $\Delta$ Fitness between novel genotypes and their wild-type counterparts were no longer significantly different from 0, except for one genotype, I–OX (Fig. 4.B, SI Appendix, Table S5). For all genotypes, $\Delta$ Fitness had diminished in magnitude after 25 generations, indicating an improvement of the negative effects after disrupting coevolution between sex chromosomes at the start of the evolution experiment, which could indicate a strong selective pressure to return the sex chromosomes to their original state.

We repeated the two phenotypic assays that had showed to either increase male fitness or decrease offspring fitness. We did no longer find an effect of genotypes on offspring egg-adult survival ($P = 0.42$, SI Appendix, Fig. S7.A and Table S4) or sperm competition ($P = 0.22$, SI Appendix, Fig. S7.B and Table S4). There was also no significant effect of genotypes on sex ratio ($P = 0.22$, SI Appendix, Fig. S7.C and Table S4), though there were significant difference in sex ratio between novel Y and wild-type (Turkey HSD: Novel Y–wild-type: $P = 0.01$).


# Conclusion

We wanted to empirically test if the sex chromosomes in D. melanogaster coevolve and found evidence of intergenomic conflict between the sex chromosomes. We confirmed previous results from (6) that variation on the Y-chromosome is an important contributor to male fitness. By swapping sex chromosomes between different genetic backgrounds and decoupling the interaction between the sex chromosomes, we were able to establish that this interaction was epistatic and independent of any interaction with the autosomes. We hypothesised that the epistatic interaction was due to intergenomic conflict and as such should be sexually antagonistic. We found that the increase in male fitness came at the cost of offspring egg-to-adult viability. After 25 generations of experimental evolution, we found no difference in almost all novel genotypes compared to the wild-type populations. This indicates a count-adaptation of the X-chromosome to the negative female fitness effects of the Y-chromosome. While we found no effect of interaction with the autosomes at the start of the experiment, we cannot exclude that the counter-adaptation we observed during the evolution experiment can be exclusively attributed to the X-chromosome. These new insights into the interactions between the sex chromosomes may have broader consequences in furthering our understanding of how genetic and phenotypic divergence arises between populations and how this can contribute to speciation events.


# Materials and Methods

## *Drosophila* husbandry and fly stocks

The five outbred laboratory-adapted wild-type populations used were: (1) Dahomey, Africa, tropical (23); (2) Innisfail, Oceania, tropical; (3) LHM, North America, mediterranean (35); (4) Odder, Europa, temperate (36); (5) Tasmania, Oceania, temperate. To eliminate any environmental difference between the five wild-type population all were maintained under the standard LHM culturing protocol (SI Materials and Methods) for at lest two generations before the cross (25°C, 12-12 light-dark cycle, 60% relative humidity, cornmeal-molasses-yeast medium (35)). For assays we used an outbred LHM population homozygous for the visible brown eye (bw) genetic marker, which is a recessive allele to the wild-type red eye-colour allele.

## Male reproductive fitness

Male reproductive fitness was measured as the proportion of live offspring sired by males of the target genotype. Five adult target males were combined for two days with ten competitor LHM-bw males and 15 virgin LHM-bw females. The females were given 18 hours for oviposition in single test tubes after which they were discarded, and the test tubes were left under standard LHM conditions for 12 days. The adult offspring of the females were counted and scored for eye-colour to assess paternity. Relative fitness was calculated by dividing the fitness for each replicate by the maximum fitness across all replicates. We measured male fitness for three different assays: at generation 0 (two blocks, seven experimental replicates in each for all 25 genotypes, fitness estimates for 70 individuals per genotype), at generation 25 (two blocks, three experimental replicates per replicate population in each, fitness estimates for 60 individuals per genotype) and for the sex chromosome-autosome interactions (10 experimental replicates per genotype, fitness estimates for 50 individuals per genotype). The rest of the phenotypic assays are described in SI Materials and Methods.

## Statistical procedures

All the statistical analyses were conducted in R version $3.4.4$ (37). We used the Fitting Linear Models (lm) command with genotype as a fixed factor to test male reproductive fitness, courtship rate, egg-to-adult offspring viability, male effect on female fecundity, sex ratio, sperm competition, thorax size, and total offspring number. Significance tests were obtained using the anova command. For the significant analyses, we also carried out a post-hoc Tukey HSD test using the TukeyHSD command to estimate significant differences between the different genotype treatments. The 95% confidence intervals around $\Delta$ Fitness ($\Delta$ Fitness = $\omega$ novel genotype - $\omega$ wild-type) was calculated by bootstrap, randomly resampling 13 out of 14 data points and recalculating $\Delta$ Fitness $10,000$ times. We used the binom.test command to test the $\Delta$ Fitness estimates. To calculate male and female longevity we used the survival package (39), and the survfit function to create survival curves. Significance tests were obtained using the Survdiff command. For the evolved genotypes we used the Linear Mixed-Effects Model (lmer) command from the R package lme4 (38) with genotype nested within replicate populations. We did this for male reproductive fitness, egg-to-adult offspring survival, sperm competition and sex ratio. Significance tests were obtained using the Anova command from the R package car (40). For the sex ratio assay, we also carried out a post-hoc Tukey test using the glht command from the R package multcomp (41) to estimate significant differences between the different genotype treatments. The 95% confidence intervals around $\Delta$ Fitness ($\Delta$ Fitness = $\omega$ novel genotype - $\omega$ wild-type) was calculated by bootstrap, randomly resampling $11$ out of $12$ data points and recalculating $\Delta$ Fitness $10,000$ times.

## **Population genetic models**

We present two population genetic models, identified by the location of the compensatory locus: the Autosomal and X-linked models respectively. Generations are assumed to be discrete, and the life-cycle proceeds: (i) birth, (ii) selection on offspring survival, (iii) meiosis and mutation, (iv) selection on male mating success. In each model, both loci are biallelic. The Y-linked locus, **Y**, has wild-type $Y$  and mutant $y$ alleles with frequencies $q_Y$ and $q_y = 1 - q_Y$ respectively. The compensatory locus is denoted **A** (with alleles $A$ and $a$ and frequencies $q_A$ and $q_a = 1 - q_A$ respectively) for the autosomal model, and **X** (with alleles $X$ and $x$ and frequencies $q_X$ and $q_x = 1 - q_X$) for the X-linked model (capital letters indicate wild-type alleles, while lowercase letters indicate mutant compensatory alleles). Following standard population genetics theory, **Y** is effectively haploid with paternal inheritance, while the autosomal or X-linked compensatory loci (**A** or **X**) are diploid with bi-parental inheritance.

A mutant $y$ chromosome increases male mating success by a rate $1 + s_m$ relative to males carrying the ancestral $Y$ chromosome. ...*brief reference to experimental results?*.... Offspring survival depends on the paternal genotype at **Y** and the maternal genotype at the compensatory locus, *as might be expected if...*. Offspring sired by mutant $y$ males experience reduced survival depending on the mother’s genotype at the compensatory locus such that $[y:AA]$, $[y:Aa]$, and $[y:aa]$ matings result in relative offspring fitness expressions of $1 − so$, $1 − h_o s_o$, and $1$. Females carrying the mutant $a$ allele may incur a ’cost of compensation’ when mating with wild-type ($Y$) males: $[Y:AA]$, $[Y:Aa]$, and $[Y:aa]$ result in relative offspring fitnesses of $1$, $1 − h_c s_c$, and $1 − s_c$ respectively (see **Table X**). Similar to standard theory for compensatory evolution (e.g., **Kimura 1985, Weinrich \& Chao 2005**) each of the mutant alleles ($y$ and either $a$ or $x$) is deleterious for offspring survival in isolation, and compensation requires that the other parent also has the appropriate mutant genotype at the other locus.

We model evolutionary invasion and single coevolutionary cycles between the male-beneficial Y-linked locus and the compensatory locus (see Connallon et al. 2017 for a similar approach in the context of mito-nuclear coevolution). A bout of coevolution begins with the invasion of a single-copy mutant $y$ chromosome in a population initially fixed for the wild-type $A$ (or $X$) allele at the compensatory locus. The mutant $y$ evolves under net positive selection if the increase in male mating success outweighs the reduction in offspring survival (i.e., if $s_{\delta} = (s_m − s_o) > 0$) until it becomes fixed in the population ($q_y = 1$) or is lost ($q_y = 0$). The compensatory locus (either **A** or **X**) evolves under recurrent mutation and selection, with the population initially fixed for the wild-type $A$ allele ($q_a = 1$ for the autosomal model) or $X$ ($q_x = 1$ for the X-linked model). For simplicity we assume one-way mutation from $A$ to $a$ at a rate $u_a$, and $X$ to $x$ at a rate $u_x$ per meiosis for X-linked model respectively. When females experience a 'cost of compensation' in terms of offspring survival (i.e., when $s_c > 0$), the mutant compensatory locus will evolve under purifying selection until the mutant $y$ chromosome reaches a threshold frequency, denoted $\tilde{q}_{y,i}$ (where $i \in \{A,X\}$), at which $y$ becomes selectively favoured. A coevolutionary cycle completes when both loci become fixed for their mutant allele (i.e., $q_y = q_a = 1$ or $q_y = q_x = 1$).

All of our analytic results assume large population sizes and an equal sex ratio. To identify the conditions under which rare mutant alleles at each locus can spread during key points of a ecolutionary cycle, we performed a linear stability analysis for each model under three different scenarios: (i) invasion of mutant a $y$ chromosome into populations initially fixed for the wild-type allele at both loci (initial frequencies of $q_y = 0$, and $q_i = 0$; where $i \in \{a,x\}$); (ii) invasion of a rare mutant compensatory allele into a population fixed for $y$ ($q_y = 1$, and $q_i = 0$); and (iii) invasion of a mutant compensatory allele into a population with an arbitrary intial frequency of $y$ ($q_y = q_y$, $q_i = 0$). Mutant alleles can invade (i.e., the initial equilibrium is unstable) when the leading eigenvalue of the Jacobian of the system of recursions is greater than one ($\lambda_L > 1$; Otto & Day 2007). When the initial frequency of $y$ is arbitrary, solving the expression $\lambda_L > 1$ for $q_y$ yields the threshold frequency at which compensatory mutations become selectively favoured, ($\tilde{q}_{y,i}$). A full derivation of all models and analytic results is provided in *Appendix X in the Online Supplementary Materials*.

To complement the analytic results, we performed stochastic Wright-Fisher simulations for the Autosomal and X-linked models with population size $N$ and an equal sex ratio. We used the simulations to obtain estimates for two important properties of coevolutionary cycles predicted by our models: (i) the probability of invasion of single-copy autosomal ($\Pi_a$) and X-linked ($\Pi_x$) compensatory mutations into populations initially fixed for the mutant $y$ chromosome; and (ii) the total time to complete a single bout of coevolution between the Y-linked and compensatory loci ($T_{cycle,A}$ and $T_{cycle,X}$ for the autosomal and X-linked models respectively) under recurrent mutation, selection, and drift. Additional details and R code for the simulations can be found in *Appendix X in the Online Supplementary Materials*, and online at https://github.com/colin-olito/sexChromCoAdapt).





# Mitochondrial model excerpts for appendix (maybe)
 and M (with alleles M and m) if it is mitochondrial

, while a mitochondrial compensatory locus (M) is haploid, with homoplasmic maternal inheritance (e.g., see Frank and Hurst 1996; Connallon et al. 2017; Roze et al. 2005)

In contrast, a Mitochondrial compensatory locus is haploid, and so the fitness expressions concern matings between [Y:M], [Y:m], [y:M], and [y:m] parental genotypes. The resulting fitness expressions are identical to those described above involving homozygote mothers (see Table 1).

For brevity, we describe the  for the Autosomal model in detail, and only note differences from the X-linked and Mitochondrial models (see Table 1). 

# Choss

, weak selection ($0 < s_{\delta}, s_c < 1$)

weak selection ($0 \ll s_{\delta}, s_c > 0$) but strong population-scaled selection ($N s_m$, $N s_o$, $N s_c \ll 1$), where $N$ is the size of a Wright-Fisher population.

The appearance of a new mutant $y$ chromosome at time $t = 0$ initiates a coevolutionary cycle. The frequency of y at time t, qy,t, increases until it fixes in the population (i.e., until qy,t = 1). Following previous theory, we use the deterministic increase of qy,t, conditional on its eventual fixation to approximate the trajectory of frequency trajectory of y during a sweep (e.g., Maynard Smith 1976; Ewens 2004; see Appendix A). For a positively selected y chromosome under weak net selection ($0 \ll \delta < 1$), the expected evolutionary trajectory of the allele frequency can be approximated as

<!--
# References
\setlength{\parindent}{-0.2in}
\setlength{\leftskip}{0.2in}
\setlength{\parskip}{8pt}
\vspace*{-0.2in}
\noindent
-->